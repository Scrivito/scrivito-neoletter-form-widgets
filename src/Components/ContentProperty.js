import * as React from "react";
import * as Scrivito from "scrivito";

/** A copy of the Scrivito ContentProperty component */
export const ContentProperty = Scrivito.connect(
  ({ content, attribute, title, description }) => {
    const validationResults = Scrivito.validationResultsFor(content, attribute);
    const highestSeverity = findHighestSeverity(validationResults);
    const severityClass = highestSeverity && `scrivito_${highestSeverity}`;

    return (
      <>
        <div className={`scrivito_detail_label ${severityClass}`}>
          {severityClass && (
            <i className="scrivito_icon scrivito_icon_error"></i>
          )}
          <span>{title}</span>
        </div>
        <div className={`content_property_input ${severityClass}`}>
          <Scrivito.ContentTag content={content} attribute={attribute} />
        </div>
        <div>
          {validationResults.map((v) => (
            <div
              key={`${v.severity}${v.message}`}
              className={`scrivito_validation_notice scrivito_${v.severity}`}
            >
              <span className="scrivito_validation_message">{v.message}</span>
            </div>
          ))}
        </div>
        {description && (
          <div className="scrivito_notice_body">{description}</div>
        )}
      </>
    );
  }
);

function findHighestSeverity(validationResults) {
  const highestSeverityValidation =
    validationResults.find((v) => v.severity === "error") ||
    validationResults.find((v) => v.severity === "warning") ||
    validationResults.find((v) => v.severity === "info");
  return highestSeverityValidation?.severity;
}
