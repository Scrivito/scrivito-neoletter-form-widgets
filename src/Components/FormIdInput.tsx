import * as React from "react";
import * as Scrivito from "scrivito";

interface ValidationResultObject {
  severity?: string;
  message: string;
}
interface FormIdInputProps {
  widget: Scrivito.Widget;
  title: string;
  description: string;
  initialId: string;
  currentId: string;
  onGenerateNewId: () => void;
  onRestoreId: () => void;
  onInputChange: (newValue: string) => void;
}

export const FormIdInput = Scrivito.connect(
  ({ widget, title, description, initialId, currentId, onGenerateNewId, onRestoreId, onInputChange }: FormIdInputProps) => {
    const validationResults = Scrivito.validationResultsFor(widget, "formId");
    const highestSeverity = findHighestSeverity(validationResults);
    const severityClass = highestSeverity && `scrivito-${highestSeverity}`;
    const [value, setValue] = React.useState<string>(widget.get("formId") as string);
    const [isModified, setIsModified] = React.useState<boolean>(false);

    React.useEffect(() => {
      setValue(currentId);
      setIsModified(currentId !== initialId);
    }, [widget, initialId, currentId]);

    const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
      const newValue = e.target.value;
      setValue(newValue);
      setIsModified(newValue !== initialId);
      onInputChange(newValue);
    };

    const handleButtonClick = () => {
      if (isModified) {
        onRestoreId();
        setValue(initialId);
        setIsModified(false);
      } else {
        onGenerateNewId();
        setIsModified(true);
      }
    };

    return (
      <>
        <div className={`detail-label ${severityClass}`}>
          {severityClass && (
            <i className="scrivito-icon scrivito-icon-error"></i>
          )}
          <span>{title}</span>
        </div>
        <div className={`content-property-input ${severityClass}`}>
          <div className="input-container">
            <input
              type="text"
              value={value}
              onChange={handleChange}
              className=""
            />
            <button className="icon-button" onClick={handleButtonClick}>
              {isModified ? <i className="bi bi-arrow-counterclockwise"></i> : <i className="bi bi-plus-circle"></i>}
            </button>
          </div>
        </div>
        <div>
          {validationResults.map(v => (
            <div
              key={`${v.severity}${v.message}`}
              className={`validation-notice scrivito-${v.severity}`}>
              <span className="validation-message">{v.message}</span>
            </div>
          ))}
        </div>
        {description && (
          <div className="notice-body">{description}</div>
        )}
      </>
    );
  }
);

function findHighestSeverity(
  validationResults: ValidationResultObject[]
): string {
  const highestSeverityValidation =
    validationResults.find(v => v.severity === "error") ||
    validationResults.find(v => v.severity === "warning") ||
    validationResults.find(v => v.severity === "info");
  return highestSeverityValidation?.severity || "";
}
