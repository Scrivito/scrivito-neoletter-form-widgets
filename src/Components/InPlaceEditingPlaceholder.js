import * as React from "react";
import * as Scrivito from "scrivito";
const placeholderCss = {
  color: "rgba(64, 64, 64, 0.53)",
  display: "inline-block",
  fontSize: "13px",
  fontStyle: "normal",
  fontWeight: "normal",
  lineHeight: "28px",
  verticalAlign: "middle",
};
/** A copy of Scrivitos InPlaceEditingPlaceholder */
export const InPlaceEditingPlaceholder = ({ children, center, block }) => {
  if (!Scrivito.isInPlaceEditingActive()) return null;

  const innerSpan = <span style={placeholderCss}>{children}</span>;

  if (center) return <div className="text-center">{innerSpan}</div>;

  if (block) return <div>{innerSpan}</div>;

  return innerSpan;
};
