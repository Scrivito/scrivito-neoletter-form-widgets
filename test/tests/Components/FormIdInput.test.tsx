import React from "react";
import { render, fireEvent, cleanup } from "@testing-library/react";
import * as Scrivito from "scrivito";
import { FormIdInput } from "../../../src/Components/FormIdInput";
import renderer from "react-test-renderer";

// Mock data for validation results
const validationResults = [
  { severity: "error", message: "Error message" },
  { severity: "warning", message: "Warning message" }
];

jest.mock("scrivito", () => ({
  connect: (component: React.FC) => component,
  validationResultsFor: jest.fn(() => validationResults)
}));

describe("FormIdInput Component", () => {
  const mockWidget = {
    get: jest.fn((attr: string) => {
      if (attr === "formId") return "testFormId";
    }),
    update: jest.fn()
  } as unknown as Scrivito.Widget;

  const initialId = "initialId";
  const currentId = "currentId";
  const onGenerateNewId = jest.fn();
  const onRestoreId = jest.fn();
  const onInputChange = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    cleanup();
  });

  it("renders the FormIdInput component with validation results", () => {
    const title = "Test Title";
    const description = "Test Description";

    const { getByText, getByDisplayValue } = render(
      <FormIdInput
        widget={mockWidget}
        title={title}
        description={description}
        initialId={initialId}
        currentId={currentId}
        onGenerateNewId={onGenerateNewId}
        onRestoreId={onRestoreId}
        onInputChange={onInputChange}
      />
    );

    expect(getByText("Test Title")).toBeInTheDocument();
    expect(getByText("Test Description")).toBeInTheDocument();
    expect(getByDisplayValue("currentId")).toBeInTheDocument();

    validationResults.forEach((result) => {
      expect(getByText(result.message)).toBeInTheDocument();
    });
  });

  it("calls onInputChange when input value changes", () => {
    const title = "Test Title";
    const description = "Test Description";

    const { getByDisplayValue } = render(
      <FormIdInput
        widget={mockWidget}
        title={title}
        description={description}
        initialId={initialId}
        currentId={currentId}
        onGenerateNewId={onGenerateNewId}
        onRestoreId={onRestoreId}
        onInputChange={onInputChange}
      />
    );

    const input = getByDisplayValue("currentId") as HTMLInputElement;
    fireEvent.change(input, { target: { value: "newId" } });

    expect(onInputChange).toHaveBeenCalledWith("newId");
  });


  it("calls onRestoreId when restore button is clicked", () => {
    const modifiedId = "modifiedId";

    const { getByRole, rerender } = render(
      <FormIdInput
        widget={mockWidget}
        title="Test Title"
        description="Test Description"
        initialId={initialId}
        currentId={modifiedId}
        onGenerateNewId={onGenerateNewId}
        onRestoreId={onRestoreId}
        onInputChange={onInputChange}
      />
    );

    const button = getByRole("button");
    fireEvent.click(button);

    expect(onRestoreId).toHaveBeenCalled();

    // ensure the button icon updates correctly
    rerender(
      <FormIdInput
        widget={mockWidget}
        title="Test Title"
        description="Test Description"
        initialId={initialId}
        currentId={initialId}
        onGenerateNewId={onGenerateNewId}
        onRestoreId={onRestoreId}
        onInputChange={onInputChange}
      />
    );

    expect(button.querySelector("i.bi-plus-circle")).toBeInTheDocument();
  });
  it("calls onGenerateNewId when generate button is clicked", () => {
    const title = "Test Title";
    const description = "Test Description";

    const { getByRole } = render(
      <FormIdInput
        widget={mockWidget}
        title={title}
        description={description}
        initialId={initialId}
        currentId={initialId}
        onGenerateNewId={onGenerateNewId}
        onRestoreId={onRestoreId}
        onInputChange={onInputChange}
      />
    );

    const button = getByRole("button");
    expect(button.querySelector("i.bi-plus-circle")).toBeInTheDocument();
    fireEvent.click(button);

    expect(onGenerateNewId).toHaveBeenCalled();
  });

  it("renders correctly and matches snapshot", () => {
    const title = "Test Title";
    const description = "Test Description";

    const tree = renderer
      .create(
        <FormIdInput
          widget={mockWidget}
          title={title}
          description={description}
          initialId={initialId}
          currentId={currentId}
          onGenerateNewId={onGenerateNewId}
          onRestoreId={onRestoreId}
          onInputChange={onInputChange}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
