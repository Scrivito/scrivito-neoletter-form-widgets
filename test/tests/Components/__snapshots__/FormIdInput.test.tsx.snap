// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormIdInput Component renders correctly and matches snapshot 1`] = `
[
  <div
    className="detail-label scrivito-error"
  >
    <i
      className="scrivito-icon scrivito-icon-error"
    />
    <span>
      Test Title
    </span>
  </div>,
  <div
    className="content-property-input scrivito-error"
  >
    <div
      className="input-container"
    >
      <input
        className=""
        onChange={[Function]}
        type="text"
        value="testFormId"
      />
      <button
        className="icon-button"
        onClick={[Function]}
      >
        <i
          className="bi bi-plus-circle"
        />
      </button>
    </div>
  </div>,
  <div>
    <div
      className="validation-notice scrivito-error"
    >
      <span
        className="validation-message"
      >
        Error message
      </span>
    </div>
    <div
      className="validation-notice scrivito-warning"
    >
      <span
        className="validation-message"
      >
        Warning message
      </span>
    </div>
  </div>,
  <div
    className="notice-body"
  >
    Test Description
  </div>,
]
`;
